feat(react-18-toast): toast with React Portal implementation

TimKolberger opened the pull request:
<!---
Thanks for creating a Pull Request 💖!

Please read the following before submitting:
- PRs that adds new external dependencies might take a while to review.
- Keep your PR as small as possible.
- Limit your PR to one type (docs, feature, refactoring, ci, or bugfix)
-->

Closes #5795 

## 📝 Description

The toast package breaks with React 18.

## ⛳️ Current behavior (updates)

It uses `ReactDOM.render` to create a new React root.

## 🚀 New behavior

~~Use `ReactDOM.createRoot` if available to support React 18 without breaking React version < 18.~~

A new introduced ToastProvider renders the toasts in a Portal. The ToastProvider is rendered by the ChakraProvider.

## 💣 Is this a breaking change (Yes/No):

Yes

* for `@chakra-ui/toast`: `createStandaloneToast`
* `@chakra-ui/react` follows

### Breaking changes for `createStandaloneToast`

With React <17 Chakra UI was able to render the toast container DOM element for
you. To allow Chakra UI to support all React versions >=16.8 you need to render
the DOM element in your application code.

#### before

```ts
import { createStandaloneToast } from "@chakra-ui/toast"
const toast = createStandaloneToast()
toast({ title: "Chakra UI" })
```

#### with React 17

```tsx
import { createStandaloneToast } from "@chakra-ui/toast"
import * as ReactDOM from "react-dom"
const toastContainerId = "chakra-toast-portal"
const toastContainerElement = document.createElement("div")
window.document.body.append(toastContainerElement)
const { ToastContainer, toast } = createStandaloneToast()
ReactDOM.render(<ToastContainer />, toastContainerElement)
toast({ title: "Chakra UI" })
```

#### with React 18

```tsx
import { createStandaloneToast } from "@chakra-ui/toast"
import * as ReactDOM from "react-dom/client"
const toastContainerId = "chakra-toast-portal"
const toastContainerElement = document.createElement("div")
window.document.body.append(toastContainerElement)
const { ToastContainer, toast } = createStandaloneToast()
const root = ReactDOM.createRoot(toastContainerElement)
root.render(<ToastContainer />)
toast({ title: "Chakra UI" })
```

## 📝 Additional Information

~~This is a hotfix implementation to support React 18. This PR should be followed by an updated implementation which uses `React.createPortal`~~

## TODOs

- [x] Split in multiple PRs in Toast Update and React 18 related changes
- [x] Create Umbrella PR for React 18 Updates
- [x] Update repo and test suite to use React 18

TimKolberger commented on packages/toast/src/toast.class.tsx:
```diff
@@ -1,6 +1,6 @@
 import { isBrowser } from "@chakra-ui/utils"
 import * as React from "react"
-import { render } from "react-dom"
+import * as ReactDom from "react-dom"
```
This won't work, because we need to `import { createRoot } from 'react-dom/client'`
But then it is React 18 specific and we cannot support React 17 🤔 

I think we need to move to a toast implementation with `createPortal` immediately to support React 18.

segunadebayo replied:
That's true. Let's chat about this in the meeting today

TimKolberger commented on .changeset/new-rivers-fail.md:
```diff
@@ -0,0 +1,54 @@
+import { createStandaloneToast } from "@chakra-ui/toast"
+import * as ReactDOM from "react-dom/client"
+
+const toastContainerId = "chakra-toast-portal"
+const toastContainerElement = document.createElement("div")
+window.document.body.append(toastContainerElement)
+
+const { ToastContainer, toast } = createStandaloneToast()
+const root = ReactDOM.createRoot(toastContainerElement)
+root.render(<ToastContainer />)
```
This change is targeting createStandaloneToast [see the current docs](https://chakra-ui.com/docs/components/feedback/toast#standalone-toasts) which allows you to create toasts from outside of your React root. E.g. some vanilla JS code. In this case chakra ui created the React root for you, but with the api change of React 18 I cannot think of a way how we could support both versions at the same time.

The most common toast function `useToast` did not change.

stefee replied:
I see, I had misread and thought that it would still inherit the portal from the provider, but I can see now that the `ToastContainer` renders its own standalone provider _and_ consumer.

One use case for `createStandaloneToast` currently is for error handling on a promise rejection within the context of a React application - e.g. failed fetch or error inside an event handler. Is that an "incorrect" usage for this?

The problem I see is when to mount/unmount the ToastContainer in the React lifecycle. If you call `createStandaloneToast` inside a callback then there is no cleanup method so you might as well just use the regular toast method - and perhaps that is the answer in this case.

TimKolberger replied:
> One use case for createStandaloneToast currently is for error handling on a promise rejection within the context of a React application - e.g. failed fetch or error inside an event handler. Is that an "incorrect" usage for this?

The use case seems good to me!

I think it depends on where createStandaloneToast is called.

I call createStandaloneToast in its own file and reuse the toast function everywhere in my code where I don't have access to useToast. 

But if it's called in the React lifecycle it's best to cleanup the created Dom nodes.

Global state vs. cleanup 🤷

stefee replied:
If you mutate DOM in the global scope would that cause some weirdness during dev if you’re hot reloading the code?

stefee replied:
I’m really questioning the value of this helper function at all as part of a React library.

TimKolberger replied:
> If you mutate DOM in the global scope would that cause some weirdness during dev if you’re hot reloading the code?

I am not an expert in hot reloading, but it haven't seen any Github issues regarding this.

gaearon commented:
I’m probably missing something about this solution…

The way portals are supposed to work is that there’s just one root in your app. You don’t need a separate root for a portal — that defeats the purpose of portals. A portal should be rendered right within your normal application tree.

Can you help me understand why the updated API still forces the user to create a separate root? Thanks.

TimKolberger commented:
Sure, thanks for asking!

The toast package support two use cases. The common one is to show a toast in your React application:

```tsx
import * as ReactDOM from 'react-dom/client'
import { ChakraProvider, useToast } from '@chakra-ui/react'

const App = () => {
  const toast = useToast()
  return <button onClick={() => toast({ title: 'Success' })}
}

ReactDOM
  .createRoot(someRootElement)
  .render(
    <ChakraProvider>
      <App />
    </ChakraProvider>
  )
```

>The way portals are supposed to work is that there’s just one root in your app. You don’t need a separate root for a portal — that defeats the purpose of portals. A portal should be rendered right within your normal application tree.

The previous implementation used a separate React root to render the toasts using `createStandaloneToast`, to which I'll get back to later on. 

In this PR I replaced the additional React root with a ToastProvider, which is rendered in the ChakraProvider and uses [Chakra UI's `Portal` component](https://chakra-ui.com/docs/components/other/portal) (which uses `React.createPortal`) to render the toasts.

**There are no breaking changes to the useToast API, and the user does not need to create an additional React root to render toasts.**

>Can you help me understand why the updated API still forces the user to create a separate root? Thanks.

Now to the confusing part: [Some time ago I introduced a new helper function `createStandaloneToast`](https://github.com/chakra-ui/chakra-ui/pull/1723) to allow showing toasts from outside of your React application.
The use case here is to "just show a toast" from a vanilla JS section of your codebase e.g. an API adapter, global error handler, etc:

```ts
import { createStandaloneToast } from "@chakra/react"

const toast = createStandaloneToast()

const fetchUsers = async () => {
  await fetch() // ...
  toast({
    title: "Yay",
    status: "success"
  })
}
```

`createStandaloneToast` created an additional React root for the user with `ReactDOM.render` - the word `standalone` in the name was designed to indicate that, but naming is hard.

With the updated root API of React 18 I can't think of a way to support React v17 and v18 at the same time, because the import path changed from "react-dom" to "react-dom/client".

If the user still has the mentioned use case, they now need to create the additional React root in their application code.
`createStandaloneToast` was now reimplemented to return the `ToastContainer` and the `toast` function.
The `ToastContainer` is meant to be rendered by React - in an additional React root or in the main applications React root.

This code example is a drop in replacement for the previous implementation of `createStandaloneToast`:

```tsx
import { createStandaloneToast } from "@chakra-ui/toast"
import * as ReactDOM from "react-dom"

const toastContainerElement = document.createElement("div")
window.document.body.append(toastContainerElement)

const { ToastContainer, toast } = createStandaloneToast()
ReactDOM.render(<ToastContainer />, toastContainerElement)

toast({ title: "Chakra UI" })
```

Let me know if this makes sense to you - I am always open for suggestions and improvements.

stefee commented:
The reason why stand-alone needs its own React root is because it consumes values from the Chakra context, so it needs to mount its own version of the Chakra context provider since React context values can’t be referenced outside of components/hooks. Is that correct?

gaearon commented:
>The ToastContainer is meant to be rendered by React - in an additional React root or in the main applications React root.

Oh okay. So the canonical usage is probably to place `<ToastContainer />` into the app tree rather than create extra roots. I missed that.

stefee commented:
A code example which shows stand-alone and regular toast used in the same app might help to clarify - I’m struggling to imagine what it looks like.

TimKolberger commented:
> So the canonical usage is probably to place <ToastContainer /> into the app tree rather than create extra roots

That's a great idea! I'll update the usage docs for `createStandaloneToast` accordingly.
Thank you for the valuable input @gaearon 💖 

[I created a CodesandBox with an example implementation](https://codesandbox.io/s/chakra-ui-2-0-0-next-createstandalonetoast-8xjzck?file=/src/index.tsx:394-795) - this is the tl;dr:

```tsx
const App = () => (
  <ChakraProvider>
    <MyToastButton />
  </ChakraProvider>
);

// create the standalone toast before rendering your App
const { ToastContainer, toast } = createStandaloneToast();

const rootElement = document.getElementById("root");
ReactDOM.createRoot(rootElement).render(
  <>
    <App />
    {/* render the ToastContainer in your React root */}
    <ToastContainer />
  </>
);
```

@stefee 
> The reason why createStandaloneToast needs its own React root is because Toast consumes values from the Chakra context, so it needs to mount its own version of the Chakra context provider since React context values can’t be referenced outside of components/hooks. Is that correct?

`createStandaloneToast` composes all necessary Chakra UI providers, like `ThemeProvider` and `ToastProvider`.
 With this in place the `ToastContainer` does not need to be a child of your applications `ChakraProvider` - it is independent.
If you are curious, here is the glue code which allow the `toast` function of `createStandaloneToast` to get access to all necessary functions:

https://github.com/chakra-ui/chakra-ui/blob/d255f1be0f66fbb051d7234f0bb038615e723268/packages/toast/src/toast.provider.tsx#L100-L101

https://github.com/chakra-ui/chakra-ui/blob/d255f1be0f66fbb051d7234f0bb038615e723268/packages/toast/src/create-standalone-toast.tsx#L68-L69

stefee commented:
This makes a lot of sense! Thanks for bearing with me on that.

stefee commented:
@TimKolberger 
I’m going to test this in a Next.js project. I think the ToastContainer should go in either `_app` or `_document` but not sure which. It might help to add a specific note about Next.js to the upgrade guide so I can report back when I figured out which is the best approach?

stefee commented:
RE: Next.js usage, either of the following will work for the `_app.js` or `_app.tsx` file.

This works ✅ 

```js
import { ChakraProvider, createStandaloneToast } from "@chakra-ui/react";

const { toast, ToastContainer } = createStandaloneToast();

export const standaloneToast = toast;

function MyApp({ Component, pageProps }) {
  return (
    <>
      <ChakraProvider>
        <Component {...pageProps} />
        <ToastContainer />
      </ChakraProvider>
    </>
  );
}

export default MyApp;
```

This also works ✅ 

```js
import { ChakraProvider, createStandaloneToast } from "@chakra-ui/react";

const { toast, ToastContainer } = createStandaloneToast();

export const standaloneToast = toast;

function MyApp({ Component, pageProps }) {
  return (
    <ChakraProvider>
      <Component {...pageProps} />
      <ToastContainer />
    </ChakraProvider>
  );
}

export default MyApp;
```

Adding to `_document.js` or `_document.tsx` does not work 👎🏻
